From 0130a3f20814ffdc6165fda25ce094cee4c5d324 Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Mon, 30 Jul 2012 19:59:36 +0000
Subject: [PATCH] [CONFIGURATION-501] XMLPropertyListConfiguration now logs a
 warning for an invalid date property rather than swallowing the parsing
 exception.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/configuration/trunk@1367253 13f79535-47bb-0310-9956-ffa450edef68
---
 src/changes/changes.xml                       |  4 +++
 .../plist/XMLPropertyListConfiguration.java   | 16 ++++++++--
 .../TestXMLPropertyListConfiguration.java     | 14 ++++++++
 .../resources/test_invalid_date.plist.xml     | 32 +++++++++++++++++++
 4 files changed, 63 insertions(+), 3 deletions(-)
 create mode 100644 src/test/resources/test_invalid_date.plist.xml

diff --git a/src/changes/changes.xml b/src/changes/changes.xml
index 17f6fee64c..b358e020d9 100644
--- a/src/changes/changes.xml
+++ b/src/changes/changes.xml
@@ -27,6 +27,10 @@
   <body>
     <release version="1.9" date="in SVN"
       description="TBD">
+      <action dev="oheger" type="add" issue="CONFIGURATION-501">
+        XMLPropertyListConfiguration no longer swallows exception caused by
+        invalid date properties. Now a warning message is logged.
+      </action>
       <action dev="oheger" type="fix" issue="CONFIGURATION-495">
         List properties can now be set correctly on a HierarchicalConfiguration
         if delimiter parsing is disabled.
diff --git a/src/main/java/org/apache/commons/configuration/plist/XMLPropertyListConfiguration.java b/src/main/java/org/apache/commons/configuration/plist/XMLPropertyListConfiguration.java
index 84ed8bf25d..141ffc0c0d 100644
--- a/src/main/java/org/apache/commons/configuration/plist/XMLPropertyListConfiguration.java
+++ b/src/main/java/org/apache/commons/configuration/plist/XMLPropertyListConfiguration.java
@@ -459,7 +459,7 @@ private static Map<String, Object> transformMap(Map<?, ?> src)
     /**
      * SAX Handler to build the configuration nodes while the document is being parsed.
      */
-    private static class XMLPropertyListHandler extends DefaultHandler
+    private class XMLPropertyListHandler extends DefaultHandler
     {
         /** The buffer containing the text node being read */
         private StringBuilder buffer = new StringBuilder();
@@ -578,7 +578,15 @@ else if ("data".equals(qName))
                 }
                 else if ("date".equals(qName))
                 {
-                    ((PListNode) peek()).addDateValue(buffer.toString());
+                    try
+                    {
+                        ((PListNode) peek()).addDateValue(buffer.toString());
+                    }
+                    catch (IllegalArgumentException iex)
+                    {
+                        getLogger().warn(
+                                "Ignoring invalid date property " + buffer);
+                    }
                 }
                 else if ("array".equals(qName))
                 {
@@ -660,6 +668,7 @@ else if (getValue() instanceof Collection)
          * Parse the specified string as a date and add it to the values of the node.
          *
          * @param value the value to be added
+         * @throws IllegalArgumentException if the date string cannot be parsed
          */
         public void addDateValue(String value)
         {
@@ -684,7 +693,8 @@ public void addDateValue(String value)
             }
             catch (ParseException e)
             {
-                // ignore
+                throw new IllegalArgumentException(String.format(
+                        "'%s' cannot be parsed to a date!", value), e);
             }
         }
 
diff --git a/src/test/java/org/apache/commons/configuration/plist/TestXMLPropertyListConfiguration.java b/src/test/java/org/apache/commons/configuration/plist/TestXMLPropertyListConfiguration.java
index c8f288d332..cc8d899213 100644
--- a/src/test/java/org/apache/commons/configuration/plist/TestXMLPropertyListConfiguration.java
+++ b/src/test/java/org/apache/commons/configuration/plist/TestXMLPropertyListConfiguration.java
@@ -392,4 +392,18 @@ public void testLoadNoDictConstr() throws ConfigurationException
                 ConfigurationAssert.getTestFile("test2.plist.xml"));
         assertFalse("Configuration is empty", plist.isEmpty());
     }
+
+    /**
+     * Tests a configuration file which contains an invalid date property value.
+     * This test is related to CONFIGURATION-501.
+     */
+    @Test
+    public void testSetDatePropertyInvalid() throws ConfigurationException
+    {
+        config.clear();
+        config.setFile(ConfigurationAssert.getTestFile("test_invalid_date.plist.xml"));
+        config.load();
+        assertEquals("'string' property", "value1", config.getString("string"));
+        assertFalse("Date property was loaded", config.containsKey("date"));
+    }
 }
diff --git a/src/test/resources/test_invalid_date.plist.xml b/src/test/resources/test_invalid_date.plist.xml
new file mode 100644
index 0000000000..1d787fbf64
--- /dev/null
+++ b/src/test/resources/test_invalid_date.plist.xml
@@ -0,0 +1,32 @@
+<?xml version="1.0"?>
+<!DOCTYPE plist SYSTEM "file://localhost/System/Library/DTDs/PropertyList.dtd">
+<!--
+   Licensed to the Apache Software Foundation (ASF) under one or more
+   contributor license agreements.  See the NOTICE file distributed with
+   this work for additional information regarding copyright ownership.
+   The ASF licenses this file to You under the Apache License, Version 2.0
+   (the "License"); you may not use this file except in compliance with
+   the License.  You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+-->
+<!-- Test configuration file with an invalid date property.
+     $Id$
+-->
+<plist version="1.0">
+    <dict>
+
+        <key>string</key>
+        <string>value1</string>
+
+        <key>date</key>
+        <date>not a valid date</date>
+
+    </dict>
+</plist>
